executor/insert_common.go

```diff
@@ -215,18 +231,131 @@ func insertRows(ctx context.Context, base insertCommon) (err error) {
 		}
 		rows = append(rows, row)
 		if batchInsert && e.rowCount%uint64(batchSize) == 0 {
+			// Before batch insert, should fill the batch allocated autoIDs.
+			rows, err = e.autoIDAllocN(ctx, rows)
+			if err != nil {
+				return err
+			}
+			e.cache = e.cache[:0]
 			if err = base.exec(ctx, rows); err != nil {
 				return err
 			}
 			rows = rows[:0]
 			if err = e.doBatchInsert(ctx); err != nil {
 				return err
 			}
+			// Reset colIdx cause rows has been cut.
+			e.rowIdx = -1
 		}
 	}
+	// Fill the batch allocated autoIDs.
+	rows, err = e.autoIDAllocN(ctx, rows)
+	if err != nil {
+		return err
+	}
 	return base.exec(ctx, rows)
 }
 
+func (e *InsertValues) autoIDAllocN(ctx context.Context, rows [][]types.Datum) ([][]types.Datum, error) {
+	retryInfo := e.ctx.GetSessionVars().RetryInfo
+	length := len(e.cache)
+	for i := 0; i < length; i++ {
+		switch e.cache[i].kind {
+		case AutoIDNull:
```

Comment:
maybe we could make it more clear by putting `AutoIDNull` and `AutoIDRebase` logic into different functions

Reply:
Good idea!